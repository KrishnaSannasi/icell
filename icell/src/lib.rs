#![cfg_attr(not(feature = "std"), no_std)]

#[macro_export]
macro_rules! write_all {
    ($owner:expr => $($cells:expr),*) => {{
        $crate::write_all!(@destruct [()] [$(($cells))*] $owner.write_all($crate::hlist!($(&$cells as &$crate::ICell<_, _>),*)))
    }};
    (@destruct [$($rest:tt)*] [$first:tt $($cells:tt)*] $value:expr) => {
        match $value {
            $crate::hlist::Cons(value, rest) => $crate::write_all!(@destruct [$($rest)* , value] [$($cells)*] rest)
        }
    };
    (@destruct [()] [] $value:expr) => {
        match $value { $crate::hlist::Nil => () }
    };
    (@destruct [(), $($rest:tt)*] [] $value:expr) => {
        match $value { $crate::hlist::Nil => ($($rest)*) }
    };
}

mod core;
pub use self::core::{ICell, Identifier, Owner, Transparent};

#[doc(hidden)]
pub mod hlist;

pub mod generative;
pub mod immovable;
pub mod runtime;
pub mod typeid;
#[cfg(feature = "std")]
pub mod typeid_tl;
