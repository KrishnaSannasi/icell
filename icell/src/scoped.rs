use core::marker::PhantomData;

pub use crate::make_scoped as owner;

#[macro_export]
macro_rules! make_scoped {
    ($name:ident) => {
        let $name = $crate::scoped::ScopedId::new();
        let enforce_unique_lifetime = $crate::scoped::EnforceUniqueLifetime(&$name);
        let $name = unsafe { $crate::scoped::Scoped::create_new_scoped($name) };
        let mut $name = $crate::Owner::new($name);
    };
}

pub fn with<F: FnOnce(Owner) -> R, R>(f: F) -> R {
    Scoped::with(f)
}

struct Invariant<'a>(&'a mut &'a ());

pub type Owner<'a> = crate::Owner<Scoped<'a>>;
pub type ICell<'a, T> = crate::ICell<ScopedId<'a>, T>;

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ScopedId<'a>(PhantomData<Invariant<'a>>);
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Scoped<'a>(ScopedId<'a>);

#[doc(hidden)]
pub struct EnforceUniqueLifetime<'a>(pub &'a ScopedId<'a>);

impl Drop for EnforceUniqueLifetime<'_> {
    #[inline(always)]
    fn drop(&mut self) {}
}

impl<'a> ScopedId<'a> {
    #[inline(always)]
    pub const fn new() -> Self {
        Self(PhantomData)
    }
}

impl<'a> Scoped<'a> {
    #[doc(hidden)]
    #[inline(always)]
    pub const unsafe fn create_new_scoped(id: ScopedId<'a>) -> Self {
        Self(id)
    }

    pub fn with<R, F: FnOnce(Owner) -> R>(f: F) -> R {
        unsafe { f(Owner::new(Self::create_new_scoped(ScopedId::new()))) }
    }
}

unsafe impl crate::Transparent for ScopedId<'_> {}
unsafe impl<'a> crate::Identifier for Scoped<'a> {
    type Id = ScopedId<'a>;

    fn id(&self) -> Self::Id {
        self.0
    }

    fn check_id(&self, _id: &Self::Id) -> bool {
        true
    }
}
