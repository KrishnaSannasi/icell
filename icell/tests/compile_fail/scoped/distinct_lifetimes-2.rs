use icell::scoped::{self, ICell};

fn main() {
    scoped::with(|foo| {
        let value = ICell::new(10);

        scoped::with(|bar| {
            foo.read(&value);
            bar.read(&value);
        });
    });
}
