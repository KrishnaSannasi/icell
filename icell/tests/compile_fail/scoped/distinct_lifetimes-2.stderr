error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
  --> $DIR/distinct_lifetimes-2.rs:8:17
   |
8  |             foo.read(&value);
   |                 ^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 7:22...
  --> $DIR/distinct_lifetimes-2.rs:7:22
   |
7  |           scoped::with(|bar| {
   |  ______________________^
8  | |             foo.read(&value);
9  | |             bar.read(&value);
10 | |         });
   | |_________^
note: ...but the lifetime must also be valid for the anonymous lifetime #2 defined on the body at 4:18...
  --> $DIR/distinct_lifetimes-2.rs:4:18
   |
4  |       scoped::with(|foo| {
   |  __________________^
5  | |         let value = ICell::new(10);
6  | |
7  | |         scoped::with(|bar| {
...  |
10 | |         });
11 | |     });
   | |_____^
note: ...so that the types are compatible
  --> $DIR/distinct_lifetimes-2.rs:9:17
   |
9  |             bar.read(&value);
   |                 ^^^^
   = note: expected  `icell::core::Identifier`
              found  `icell::core::Identifier`
